Redirections

asaunier opened the pull request:
This PR:
* uses the new API info service to get the slug of partial document URLs (without lang or slug) => fixes #568
* handles merged/redirecting documents => fixes #190

asaunier commented:
Anyone available for reviewing this PR? :)

tsauerwein commented:
I will take a look.

asaunier commented:
Thanks! Feel free to merge the PR if it looks good to you.

tsauerwein commented on c2corg_ui/views/document.py:
```diff
@@ -432,12 +441,17 @@ def _redirect_to_full_url(self):
-        raise HTTPMovedPermanently(location=location)
+        if 'redirects_to' in data:
+            if lang not in data['available_langs']:
+                lang = self._get_preferred_lang()
```
Instead of using `prefered_lang`, you would have to select the *best* lang in `data['available_langs']`. Otherwise you might do a redirect to a locale that does not exist (and at this point you already know that it does not exist).

asaunier replied:
What is "the best lang in data['available_langs']"?
On the UI side we don't know what lang to choose among ``available_langs`` if ``available_langs`` does not contain the requested ``lang``. In that case, I suggest to fall back on the user's perferred lang and let the API deal with it (return the "best lang" if the preferred lang is not available).
Doesn't it make sense?

tsauerwein replied:
If you call `/waypoints/1234/fr` and there is no locale for `fr` for waypoint `1234`, you will get a `404`.

> What is "the best lang in data['available_langs']"?

https://github.com/c2corg/v6_api/blob/b32316f7ee1a9db28107e20c9678c87a7c235468/c2corg_api/views/__init__.py#L168-L175

asaunier replied:
OK, that makes sense to do as you suggest.

asaunier replied:
When the API returns ``redirects_to`` it also provides the available langs. So it's easy to figure the best lang out as in your link.

But what should we do when the ``lang`` param is not provided (URL such as ``waypoints/123456``)? We don't know what langs are available for this document. Is there a way to ask the API for the available langs?

tsauerwein replied:
> But what should we do when the lang param is not provided (URL such as waypoints/123456)? We don't know what langs are available for this document. Is there a way to ask the API for the available langs?

Isn't this PR all about handling this? You are doing a request to the "document info" service with the interface language. If a locale in that language does not exist, you get the *best* locale.

tsauerwein commented on c2corg_ui/views/document.py:
```diff
@@ -192,6 +192,12 @@ def _get_document(self, id, lang=None, old_api_cache_key=None):
+        # Manage merged documents (redirecting to another document)
+        if 'redirects_to' in document:
+            if lang is None or lang not in document['available_langs']:
+                lang = self._get_preferred_lang()
```
Same comment as above.

tsauerwein commented on c2corg_ui/views/document.py:
```diff
@@ -413,20 +421,43 @@ def _get_api_cache_key_from_etag(self, etag):
+        else:
+            location = self.request.route_url(
+                self._API_ROUTE + '_view', id=id, lang=lang, slug=slug)
+        raise HTTPMovedPermanently(location=location)
```
It's better if we use `HTTPFound` (302) instead of `HTTPMovedPermanently` (301). The browser is very eager in caching 301 redirects, it will *never* do a request again and directly use the redirected URL. I will also change it in the API.

asaunier replied:
> It's better if we use HTTPFound (302) instead of HTTPMovedPermanently (301).

The goal of using 301 is to make it clear to search engines that the redirected URL should be used instead of the original URL ("permanent redirection"). Not sure 302 would be understood the same way by search engines.

https://moz.com/learn/seo/http-status-codes says that
> Use 301 redirects rather than 302 redirects when redirecting URLs on a site to ensure that link juice (ranking power) is passed between the redirecting web pages.

and https://moz.com/learn/seo/redirection
> 301, "Moved Permanently"—recommended for SEO

tsauerwein replied:
The main problem with using `HTTPMovedPermanently (301)` is that the redirect is actually not permanent. Two examples:

When requesting `/waypoints/1234` (without lang), the redirect depends on the current interface language. Let's say the interface language is `en`, so it will redirect to `/waypoints/1234/en/...`. If the interface language is now changed to `fr`, the browser will directly redirect to `/waypoints/1234/en/...`.

Second problem: If `redirects_to` is set for a document, a permanent redirect to the linked document will be made. But `redirects_to` can be changed! So even if you change `redirects_to`, browsers will still redirect to the old linked document, because they think the redirect is permanent (while it is only temporary!).

asaunier replied:
> When requesting /waypoints/1234 (without lang), the redirect depends on the current interface language.

OK, makes sense. So we need a 302 when the lang is missing and 301 if it is provided :)

> Second problem: If redirects_to is set for a document, a permanent redirect to the linked document will be made. But redirects_to can be changed!

Why would it change? This case has never happened in v5 since 2007.

My concern is that we need to make sure that redirecting documents are not referenced by search engines and that the target document is always taken into accound instead. It seems we need the 301 for this, doesn't it?

tsauerwein replied:
> Why would it change? This case has never happened in v5 since 2007.

How did you check this? Our model does not prevent that `redirects_to` can not be changed anymore if it has been set once.

> My concern is that we need to make sure that redirecting documents are not referenced by search engines and that the target document is always taken into accound instead. It seems we need the 301 for this, doesn't it?

The redirected page will also be indexed with 302. This graphic explains it quite well: https://seohackercdn-seohacker.netdna-ssl.com/wp-content/uploads/2011/10/301-and-302-Redirect-small2.png

asaunier replied:
> How did you check this? Our model does not prevent that redirects_to can not be changed anymore if it has been set once.

Because merging documents is a moderator-only feature, undoing a merge requires an hand-change in the DB in v5 and I'm pretty sure such a operation has never been requested (there are very few people that have ssh access to the server).

Of course documents might be merged by error and rollback would then be necessary. But this will happen very rarely (this is not a public feature). And the page ranking issue is much more important.

> The redirected page will also be indexed with 302. This graphic explains it quite well: https://seohackercdn-seohacker.netdna-ssl.com/wp-content/uploads/2011/10/301-and-302-Redirect-small2.png

Well, the bottom of the graphic explains that 
<img width="477" alt="capture d ecran 2016-08-31 a 17 18 14" src="https://cloud.githubusercontent.com/assets/1192331/18134725/aba444f6-6f9f-11e6-8358-3839317a7a93.png">

In order to actually transfer the indexing of the old page to the new one, a 301 is recommended.

tsauerwein replied:
> In order to actually transfer the indexing of the old page to the new one, a 301 is recommended.

If it really is a permanent redirect, yes. I am worried that accidentally a redirect is set to a wrong document. And then you have a hard undoing the (not so) permanent redirect.

asaunier commented:
I have updated the PR so that:
* missing or unavailable langs are better managed
* a 302 (temporary) redirection is made for URL not providing the lang.

When redirecting from a URL providing both the document id and lang, I have preferred to stick to a 301 (permanent) redirection because:
* it's critical that the documents have a good page ranking in google and other search engines. Which means that only one document is indexed for a given WP/route/...
* I agree that the 301 makes it more difficult to fix when 2 documents have been merged by error but I don't think it has ever happened with the v5 (this is a moderator only feature). Same problem actually than if a moderator had dropped the document by mistake.
For me the first point is more important than the second one.

asaunier commented:
By the way, merging documents is not only about flagging one as redirecting to the other, it should also transfer all the associations of the redirecting doc to the target one (perhaps it's not done yet, as in v5). Which means that merging documents is (almost) irreversible anyway.

tsauerwein commented:
> By the way, merging documents is not only about flagging one as redirecting to the other, it should also transfer all the associations of the redirecting doc to the target one (perhaps it's not done yet, as in v5).

The API endpoint to mark a document as *redirected* is not implemented yet. This is part of the moderator tools.

> Which means that merging documents is (almost?) irreversible anyway.

We should make sure that we communicate this clearly in the UI (something like "Are you really sure...").


It would be cool to have some tests for the logic that you added in this PR. But feel free to merge.

asaunier commented:
> The API endpoint to mark a document as redirected is not implemented yet. This is part of the moderator tools.

OK, I have added an API issue to make sure we keep track of the things to do:
https://github.com/c2corg/v6_api/issues/386

> We should make sure that we communicate this clearly in the UI (something like "Are you really sure...").

Sure!

> It would be cool to have some tests for the logic that you added in this PR.

There was already some testing about the slug redirection. I have added some more with 89f89e9

tsauerwein commented:
:+1:

asaunier commented:
Thanks @tsauerwein for the review and tips!
